### PR TITLE
SQL storage

### PR DESCRIPTION
**Pull request by naveensrinivasan**
## Summary by CodeRabbit

- **New Features**
	- Introduced SQLite storage implementation, enhancing data management capabilities.
	- Added new dependencies to support advanced database operations.
	- Expanded server configuration options to allow SQLite as a storage type.
	- Implemented comprehensive unit tests for the new storage functionality.

- **Bug Fixes**
	- Improved error handling for database connection issues.

- **Documentation**
	- Updated documentation to reflect new methods and functionalities related to SQLite storage.

### CODE REVIEW

**Comment by naveensrinivasan** [comment]
@coderabbitai please review

**Comment by naveensrinivasan** [comment]
@coderabbitai please review

**Comment by naveensrinivasan** [comment]
@coderabbitai please review

File: cmd/server/server.go [hunk 1]
```diff
@@ -26,25 +26,32 @@
 	addr        string
 	StorageType string
 	StorageAddr string
+	StoragePath string
+	UseInMemory bool
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply cmd/server/server.go:30]: Ignore this, as I don't like to create a directory.

File: pkg/storages/sql.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	for _, mapping := range mappings {
+		id, err := strconv.ParseUint(mapping.Value, 10, 32)
+		if id > math.MaxUint32 {
+			return nil, fmt.Errorf("ID exceeds uint32 maximum value")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("invalid ID format for key %s: %w", mapping.Key, err)
+		}
+		ids = append(ids, uint32(id))
+	}
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/storages/sql.go]: Took care of it.

File: pkg/storages/sql.go [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+	ids := make([]uint32, len(kvNodes))
+	for i, kvNode := range kvNodes {
+		var graphNode graph.Node
+		if err := graphNode.UnmarshalJSON([]byte(kvNode.Value)); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal node data: %w", err)
+		}
+		ids[i] = graphNode.ID
+	}
+	return ids, nil
+}
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/storages/sql.go:261]: Ignore this.

File: pkg/storages/sql.go [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+			sb.WriteByte('_')
+		case '%', '_':
+			sb.WriteByte('\\')
+			sb.WriteRune(char)
+		default:
+			sb.WriteRune(char)
+		}
+	}
+	return sb.String()
+}
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/storages/sql.go:425]: Create an issue for this.

File: pkg/storages/sql.go [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+		if err := tx.Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "id"}}, // Conflict on ID
+			DoNothing: true,                          // Do nothing on conflict
+		}).Create(&cacheEntry).Error; err != nil {
+			return fmt.Errorf("failed to add node ID to cache stack: %w", err)
+		}
+
+		return nil
+	})
+}
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/storages/sql.go:151]: Ignore this.

File: pkg/storages/redis_storage.go [hunk 6]
```diff
@@ -125,21 +125,21 @@
...
-		return nil, fmt.Errorf("failed to get to_be_cached data: %w", err)
+		return nil, fmt.Errorf("failed to get %s data: %w", CacheStackKey, err)
 	}
 
 	result := make([]uint32, 0, len(data))
 	for _, item := range data {
 		id, err := strconv.ParseUint(item, 10, 32)
 		if err != nil {
-			return nil, fmt.Errorf("failed to parse item %s in to_be_cached: %w", item, err)
+			return nil, fmt.Errorf("failed to parse item %s in %s: %w", item, CacheStackKey, err)
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/storages/redis_storage.go:142]: Ignore this.

**Review by neilnaveen — APPROVED** [review]
LGTM

**Pull request merged**